### PR TITLE
feat: add product to cart from wishlist

### PR DESCRIPTION
**Pull request by AlexanderDevitsky**
## Related Issue

https://vsf.atlassian.net/browse/M2-502

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by AlexanderDevitsky** [comment]
@bartoszherba should I create unit-tests in the scope of this task?

**Comment by bartoszherba** [comment]
@AlexanderDevitsky I must put this on hold for a while, if you look at this draft https://github.com/vuestorefront/magento2/pull/1014 you will realize why. There will be multiple problems like we no longer can have business logic on the product page. I also do not like the fact that we base on the locale path as the solution is not generic, eg. if someone will add a wishlist somewhere else, or create a different path for a subaccount, or [name it] it will not work. I need to think of how it can be solved. For now we must hold this PR ;(

File: packages/theme/modules/catalog/pages/product.vue [hunk 1]
```diff
@@ -560,9 +572,48 @@ export default defineComponent({
...
+        && isProductInWishlist.value
+        && !cartError.value?.addItem
+      ) {
+        await removeItemFromWishlist({ product: product.value });
+
+        sendNotification({
+          id: Symbol('product_added_to_cart_from_wishlist_product_page'),
+          message: i18n.t(`You added ${getProductName(
+            product.value,
+          )} to your shopping cart.`) as string,
```
Comment (sethidden) [inline packages/theme/modules/catalog/pages/product.vue:590]: Please pass product name as parameter to the label. We can't possibly make thousands of translation labels for each product name

https://kazupon.github.io/vue-i18n/guide/formatting.html#named-formatting

```suggestion
          message: i18n.t(`You added productname to your shopping cart.`, { productname: getProductName(product.value)} ) as string,
```

This won't render the product name after this change obviously, so you'll need to add that to the translations in en.js, de.js like `'You added productname to your shopping cart': 'You added {productname} to your shopping cart'`.

Don't forget to do that in MyWishlist aswell.

@bartoszherba is it ok to add new translation labels like that?

Comment (bartoszherba) [reply packages/theme/modules/catalog/pages/product.vue:590]: No, the product's translation comes from Magento. We mustn't translate it on the front. Good catch, thanks.

File: packages/theme/modules/catalog/pages/product.vue [hunk 2]
```diff
@@ -560,9 +572,48 @@ export default defineComponent({
...
+          type: 'danger',
+          icon: 'cross',
+          persist: false,
+          title: 'Wishlist error',
+        });
+      }
+    };
+
+    onMounted(() => {
+      if (from.value?.path === localePath('/customer/my-wishlist')) {
```
Comment (sethidden) [inline packages/theme/modules/catalog/pages/product.vue:609]: 1. Wondering if it wouldn't be easier to just redirect to the product page with some query parameter `/p/product?from-wishlist` - that way if someone refreshes the page the "fromWishlist" information won't be lost

2. Shouldn't this be a computed?

3. If the query paramater doesn't work, it'll be easier to maybe use the route *name*, not the route path

```suggestion
      if (from.value?.name === localeRoute({ name: 'customer-my-wishlist' })) {
```

(see modules/customer/index.ts for route name)

File: packages/theme/modules/customer/pages/MyAccount/MyWishlist.vue [hunk 3]
```diff
@@ -316,6 +320,29 @@ export default defineComponent({
...
+              id: Symbol('product_added_to_cart_from_wishlist_wishlist_page'),
+              message: app.i18n.t(`You added ${productGetters.getName(
+                product,
+              )} to your shopping cart.`) as string,
+              type: 'success',
+              icon: 'check',
+              persist: false,
+              title: 'Wishlist',
+            });
+          } else if (cartError.value.addItem.message) {
```
Comment (sethidden) [inline packages/theme/modules/customer/pages/MyAccount/MyWishlist.vue:336]:
```suggestion
          } else {
```

would that work? I don't think errors can have missing messages

**Review by sethidden — CHANGES_REQUESTED** [review]

**Comment by AlexanderDevitsky** [comment]
@bartoszherba In this case, we need to extend Magento GraphQL schema by adding a 'remove_from_wishlist' field. Then implement the logic of removing a product from the wishlist on the Magento side. As I understand it, for the default theme, we cannot use additional implementations for Magento, but must use only the default implementation. Therefore, this option is not possible.
Currently, the referrer url is only used to remove configurable product when navigating to it from the wishlist.
If a product needs to be removed from the wishlist anyway, not just when added from the wishlist, then we just need to check if it's on the wishlist and using referrer url is redundant.
When https://github.com/vuestorefront/magento2/pull/1014 is finished, let me know and we will find a better solution for this.

File: packages/theme/modules/customer/pages/MyAccount/MyWishlist.vue (hunk 3 above)
Comment (AlexanderDevitsky) [reply packages/theme/modules/customer/pages/MyAccount/MyWishlist.vue:336]: @sethidden I think the message should be there anyway, so this can be removed

**Comment by bartoszherba** [comment]
@AlexanderDevitsky 
PR is merged, I am sorry for that inconvenience.

With regards to the solution:

I think that we can make it work as @sethidden suggested. I did a small POC and reworked the query build in `product.vue` to something like, for a query parameter `wishlist=true` => example `http://localhost:3000/default/p/WJ01/stellar-solar-jacket.html?wishlist=true&OTM%3D=Y29uZmlndXJhYmxlLzkzLzYw&MTQ0=Y29uZmlndXJhYmxlLzE0NC8xNjk%3D`
```
const getBaseSearchQuery = () => ({
      filter: {
        sku: {
          eq: id,
        },
      },
      configurations: Object.entries(route.value.query)
        .filter((config) => config[0] !== 'wishlist')
        .map(
          (config) => config[1],
        ) as string[],
    });
```

Now, we can trigger wishlist mode on any PDP page with this flag from any place simply by passing the query param and we are not tight to the account page. If you look on the Magento, there is something very similar, as the add-to-wishlist has its own controller.

That would be the first step.

The second step is to have generic and reusable addToWishlist logic, separated from PDP (unlike you did in your PR which is justified by how it looked at that time).
Important thing is to keep business logic out of components, especially out of the page. This is not completely done at this moment because I did not want to keep such a huge change any longer, but this is the goal, so by introducing a new feature we can do it right from the very beginning.

@sethidden @AlexanderDevitsky What do you think?

**Comment by AlexanderDevitsky** [comment]
@bartoszherba Good solution

**Review by bartoszherba — CHANGES_REQUESTED** [review]
Please, review all comments.

**Comment by AlexanderDevitsky** [comment]
@bartoszherba @sethidden I have added new changes. Please review if this is the right solution for us

File: packages/theme/modules/customer/pages/MyAccount/MyWishlist.vue [hunk 4]
```diff
@@ -183,13 +192,22 @@ export default defineComponent({
             product,
             quantity,
           });
+          afterAddingWishlistItemToCart({
+            product,
+            cartError: cartError.value.addItem,
+          });
           break;
         case 'BundleProduct':
```
Comment (bartoszherba) [inline packages/theme/modules/customer/pages/MyAccount/MyWishlist.vue:200]: could you please add GroupedProduct? I know that it is not your change but I just noticed that. Also, you can replace it with `packages/theme/modules/catalog/product/enums/ProductTypeEnum.ts`

File: packages/theme/modules/wishlist/composables/useWishlist/index.ts [hunk 5]
```diff
@@ -10,7 +10,10 @@ import type {
   UseWishlistIsInWishlistParams,
   UseWishlistLoadParams,
   UseWishlistRemoveItemParams,
+  UseWishlistAfterAddingWishlistItemToCartParams,
 } from '~/modules/wishlist/composables/useWishlist/useWishlist';
+import { useUiNotification } from '~/composables';
+import productGetters from '~/modules/catalog/product/getters/productGetters';
```
Comment (bartoszherba) [inline packages/theme/modules/wishlist/composables/useWishlist/index.ts:16]: Could you please remove that and simply `product.name` in the code? We will remove most of the getters and we avoid using getters in new code unless it is necessary, eg. when getter execute some additional logic.

**Review by bartoszherba — CHANGES_REQUESTED** [review]
@AlexanderDevitsky 
Thank you for changes, just last 2 small comments and we are done here ;)

File: packages/theme/modules/customer/pages/MyAccount/MyWishlist.vue (hunk 4 above)
Comment (AlexanderDevitsky) [reply packages/theme/modules/customer/pages/MyAccount/MyWishlist.vue]: done

File: packages/theme/modules/wishlist/composables/useWishlist/index.ts (hunk 5 above)
Comment (AlexanderDevitsky) [reply packages/theme/modules/wishlist/composables/useWishlist/index.ts]: done

**Review by bartoszherba — APPROVED** [review]

**Review by bartoszherba — APPROVED** [review]

**Review by Frodigo — APPROVED** [review]

**Review by sethidden — APPROVED** [review]

**Pull request merged**